core/src/main/java/org/apache/calcite/schema/lookup/LoadingCacheLookup.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.schema.lookup;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class can be used to cache lookups.
```

Comment:
Please say what it does, before you can say what it can be used for.
The important aspect seems to be some kind of cache expiry mechanism.

I wonder about consistency. Can you say something about this?
I am guessing this assumes (and perhaps schemas do too) that objects are only inserted, and never deleted?

Reply:
I changed the documentation accordingly:

This class is using a `LoadingCache`to speed up lookups,
 delegated to another `Lookup` instance.
 
 This class is thread safe. All entries are evicted after one minute.
 Negative matches are never cached. If a new entry
 becomes available in the associated {@code Lookup}, it's immediately
 visible outside. If an entry is deleted in the associated {@code Lookup},
 it takes one minute until it disappears if it was cached in the last minute.
 Otherwise, it disappears immediately.